lambda/lambda.ml

```diff
@@ -57,16 +61,39 @@ include (struct
     | Alloc_local, _ | _, Alloc_local -> Alloc_local
     | Alloc_heap, Alloc_heap -> Alloc_heap
 
+  let modify_heap = Modify_heap
+
+  let modify_maybe_stack : modify_mode =
+    (* CR zqian: possible to move this check to a better place? *)
+    (* idealy I don't want to do the checking here.
+       if stack allocations are disabled, then the alloc_mode which this modify_mode
+        depends on should be heap, which makes this modify_mode to be heap *)
+
+    (* one suggestion: move the check to optimize_allocation;
+      if stack_allocation not enabled, force all allocations to be heap,
+        which then propagates to all the other modes.
+       *)
```

Comment:
but if we move the check to `optimise_allocation`, there is no need to have more checks in `translcore`? Besides, `optimise_allocations` pushing all allocations to `local` even when stack allocation disabled, just doesn't feel right to me

Reply:
Putting the check here statically ensures that it is impossible for the compiler to contain any occurrences of `alloc_heap` when it shouldn't. You don't need to read any other code than this to know it is safe. Whereas I'm not even sure that putting the check in `optimise_allocation` covers all cases. For instance I think that it doesn't prevent stack allocating a `ref` via a primitive.

Reply:
Yeah that 's a good point because allocations in primitives are not registered. I still hope for a better solution but that might require too much refactoring.

Reply:
I also think the check should be here, but if you prefer you could write it as:
```ocaml
match alloc_local with
| Alloc_local -> Modify_maybe_stack
| Alloc_heap -> Modify_heap
```